tests/models/clvp/test_modeling_clvp.py

```diff
@@ -0,0 +1,598 @@
+            n_head=self.num_attention_heads,
+            bos_token_id=self.bos_token_id,
+            eos_token_id=self.eos_token_id,
+            relative_attention_num_buckets=self.relative_attention_num_buckets,
+            relative_attention_max_distance=self.relative_attention_max_distance,
+        )
+
+        return autoregressive_config
+
+    def prepare_config_and_inputs(self):
+        input_ids = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
+
+        input_mask = None
+        if self.use_input_mask:
+            input_mask = random_attention_mask([self.batch_size, self.seq_length])
+
+        if input_mask is not None:
+            batch_size, seq_length = input_mask.shape
+            rnd_start_indices = np.random.randint(1, seq_length - 1, size=(batch_size,))
+            for batch_idx, start_index in enumerate(rnd_start_indices):
+                input_mask[batch_idx, :start_index] = 1
+                input_mask[batch_idx, start_index:] = 0
+
+        autoregressive_config = self.get_config()
+
+        return autoregressive_config, input_ids, input_mask
+
+    def create_and_check_model(self, config, input_ids, attention_mask):
+        model = CLVPAutoRegressiveLMHeadModel(config).to(torch_device).eval()
+        with torch.no_grad():
+            result = model(input_ids=input_ids, attention_mask=attention_mask)
+
+        self.parent.assertEqual(result[0].shape, (self.batch_size, self.seq_length, self.vocab_size))
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        config, input_ids, attention_mask = config_and_inputs
+        inputs_dict = {
+            "input_ids": input_ids.to(torch_device),
+            "attention_mask": attention_mask.to(torch_device),
+        }
+        return config, inputs_dict
+
+
+@require_torch
+class CLVPAutoRegressiveLMHeadModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
+    all_model_classes = (CLVPAutoRegressiveLMHeadModel,) if is_torch_available() else ()
+    all_generative_model_classes = (CLVPAutoRegressiveLMHeadModel,) if is_torch_available() else ()
+
+    test_pruning = False
+
+    def setUp(self):
+        self.model_tester = CLVPAutoRegressiveLMHeadModelTester(self)
+
+    def test_model(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
+        if return_labels:
+            inputs_dict["labels"] = torch.zeros(
+                [self.model_tester.batch_size, self.model_tester.seq_length], device=torch_device
+            ).long()
+
+        return inputs_dict
+
+    @unittest.skip(
+        "This test is not valid for CLVPAutoRegressiveLMHeadModel since save_pretrained and from_pretrained can"
+        "give errors in most cases."
+    )
```

Comment:
What kind of errors do you get?

Reply:
Actually (as of now) we are not supposed to load the AutoReg model(with customized parameters) using `from_pretrained` since it will load the `CLVPConfig` and it will have the predefined params of the AutoReg model. So we will get an error because of the mismatch between the shapes of the weights. 

Is is required to make sure that the AutoReg model(with customized parameters) can be loaded using `from_pretrained`?

Reply:
> Actually (as of now) we are not supposed to load the AutoReg model(with customized parameters) using from_pretrained since it will load the CLVPConfig and it will have the predefined params of the AutoReg model. So we will get an error because of the mismatch between the shapes of the weights.

wouldn't you rather say that it will try to load the weights from the `CLVPModel`, hence causing errors in loading a state dict that doesn't match with the AutoReg model architecture ?

Reply:
Actually if we try to do `CLVPAutoRegressiveLMHeadModel.from_pretrained("susnato/clvp_dev")` it will load the correct weights with a warning that some weights are not loaded. 

But the problem comes, when we change the config file and do a `.save_pretrained(custom_dir)` and then try to load from that, `.from_pretrained(custom_dir)`. It is creating the model with the actual autoreg config(not the customized one) and then trying to load the customized weights and hence failing.

Reply:
In that case, you can try simply changing the class `config_class`, like you did in [here](https://github.com/huggingface/transformers/blob/8a3b7fae369b7f430af39a53a9f877c9f2da2c35/src/transformers/models/clvp/modeling_clvp.py#L781), with the right config class.

You should probably move [that line ](https://github.com/huggingface/transformers/blob/8a3b7fae369b7f430af39a53a9f877c9f2da2c35/src/transformers/models/clvp/modeling_clvp.py#L1092) to `CLVPModel`, i.e sthg like that:

```python
self.autoregressive_model = CLVPAutoRegressiveLMHeadModel(config.autoregressive_config)
```

Reply:
Thanks for the suggestion! BTW could you please take a look at this [comment](https://github.com/huggingface/transformers/pull/24745#discussion_r1313260917), I am asking because I am currently working on this and concerned if this compiles with the library or not.

Reply:
I fixed it, actually we needed to override the `to_json_string` method for `CLVPAutoRegressiveConfig`so that it would store the config as {"autoregressive_config": {...}} and it would update the `CLVPConfig` when we are trying to do `save_pretrained()`. 
In the previous version the autoreg config attributes would just be considered as **kwargs and not as autoreg config so this was not updating the `CLVPConfig.autoregressive_config` properly thus creating the problem.